### PR TITLE
Add VRO Console container

### PR DESCRIPTION
**Pull request by yoomlam**
## What was the problem?

With VRO's deployment to prod, we'll need ways to investigate and recover from errors in the production environment.

When an error occurs, to investigate and diagnose, we'll need to examine the processing state of the claim by looking at Camel routes and the DB contents.

Associated ticket: #502 

## How does this fix it?[^1]

Create a "VRO Console" container that connects to the DB to allow investigation of DB contents.
To keep this relatively PR small, the following will be in separate PRs:
- Connecting with Camel within the Console.
- Adding this console container to deploy scripts and GH Actions
- Tests for the new code
- Add Groovy test code coverage check
- ~Remove `LABEL` command from all `Dockerfile`s~ (SecRel uses the Dockerfile directly, so let's keep the LABEL)

## How to test this PR
- `export COMPOSE_PROFILES=debug` -- the console container only starts if the docker-compose `debug` profile is enabled.
- start VRO. Note the new `vro-console-1` container; check out the logs: `docker logs vro-console-1`.
- Attach to the container: `docker attach vro-console-1`
- Review https://groovy-lang.org/groovysh.html
- On the `groovy:000>` prompt, try the following:
```groovy
?  # display help
:show variables
claimsT.findAll().collect{ it.claimSubmissionId }
c = getAnyClaim()
// Different ways to do the same thing:
printJson c
printJson getAnyClaim()
pj c
// exit   # This will stop the container
// Instead, press Ctrl-p Ctrl-q to detach from the container without stopping it
```

If you need to restart the console container:
```
cd app/src/docker
docker-compose up -d console
docker attach vro-console-1
```

[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
[^secrel]: To check if a PR will succeed in the SecRel workflow, [test PRs in the SecRel pipeline](https://github.com/department-of-veterans-affairs/abd-vro-internal/wiki/Secure-Release-process#to-test-prs-in-the-secrel-pipeline).

### CODE REVIEW

File: console/src/docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+FROM eclipse-temurin:17-jre-alpine
+LABEL org.opencontainers.image.source=https://github.com/department-of-veterans-affairs/abd-vro
+
+RUN apk update && apk --no-cache add expat=2.4.9-r0
```
[the remark this answers is not in the document]
Comment (yoomlam) [reply console/src/docker/Dockerfile:4]: This rule doesn't recognize `apk add` is doing an install.

**Comment by yoomlam** [comment]
```bash
> docker images --filter "label=org.opencontainers.image.source=https://github.com/department-of-veterans-affairs/abd-vro"
REPOSITORY                       TAG       IMAGE ID       CREATED              SIZE
va/abd_vro-app                   latest    d08e8ede6851   About a minute ago   252MB
va/abd_vro-service-data-access   latest    6be2e6a58f08   8 minutes ago        314MB
va/abd_vro-console               latest    e21112b0a878   About an hour ago    245MB
va/abd_vro-pdfgenerator          latest    db0ac993bf56   5 hours ago          167MB
va/abd_vro-assessclaimdc7101     latest    02d863947789   5 hours ago          53.8MB
va/abd_vro-assessclaimdc6602     latest    eff60bedcf3e   5 hours ago          53.8MB
va/abd_vro-postgres              latest    e610ac34d9d2   5 hours ago          216MB
va/abd_vro-db-init               latest    bbc31c6f85b4   5 hours ago          276MB
va/abd_vro-service-ruby          latest    c16b1aa888e5   3 months ago         431MB
```

To prune VRO images:`./gradlew :app:dcPrune` or
```bash
docker system prune --filter "label=org.opencontainers.image.source=https://github.com/department-of-veterans-affairs/abd-vro"
```

File: app/src/docker/docker-compose.yml [hunk 2]
```diff
@@ -190,3 +191,29 @@ services:
     networks:
       - internal
 
+  vro-console:
+    image: va/abd_vro-console:latest
+    profiles: ["debug"]
```
Comment (yoomlam) [inline app/src/docker/docker-compose.yml:196]: The console container only starts if the docker-compose `debug` profile is enabled: `export COMPOSE_PROFILES=debug`

File: service-data-access/build.gradle [hunk 3]
```diff
@@ -98,11 +98,6 @@ task integrationTest(type: Test) {
     }
 }
 
-docker {
-    files "build/libs/${bootJar.archiveFileName.get()}", "bin/entrypoint.sh"
-    buildArgs([JAR_FILE: bootJar.archiveFileName.get(), ENTRYPOINT_FILE: "entrypoint.sh"])
-}
-
```
Comment (yoomlam) [inline service-data-access/build.gradle:105]: This common code has been moved to https://github.com/department-of-veterans-affairs/abd-vro/pull/531/files#diff-49a96e7eea8a94af862798a45174e6ac43eb4f8b4bd40759b5da63ba31ec3ef7R104-R110

File: build.gradle [hunk 4]
```diff
@@ -96,7 +96,20 @@ subprojects {
     afterEvaluate {
         if (tasks.findByName('docker')) {
             docker {
+                // Use cached image; only recreate if different
                 noCache false
+                labels(['org.opencontainers.image.source':'https://github.com/department-of-veterans-affairs/abd-vro'])
```
Comment (yoomlam) [inline build.gradle:101]: `LABEL org.opencontainers.image.source=https://github.com/department-of-veterans-affairs/abd-vro` can now be removed from `Dockerfile`s like https://github.com/department-of-veterans-affairs/abd-vro/blob/8cdc31ffafc5bfa3959c8c80336a79cdcf3b7a31/app/src/docker/Dockerfile#L2

File: console/build.gradle [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    id 'starter.std.java.library-spring-conventions'
+	// Needed for bootRun task
+	id 'org.springframework.boot'
+	// Needed for tasks docker and dockerRun
+    id 'starter.java.container-conventions'
+	id 'groovy'
+}
+
+ext {
+//	set('testcontainersVersion', "1.17.4")
```
Comment (yoomlam) [inline console/build.gradle:11]: May be uncommented in separate PR where tests will be added.

**Comment by yoomlam** [comment]
[SecRel passed](https://github.com/department-of-veterans-affairs/abd-vro-internal/actions/runs/3246027416)

**Comment by au2** [comment]
@yoomlam I do not understand what we are trying to do here based on the description. If we can connect to production with this "console" we should be able to connect to postgres container and run psql just as well.  Actually based on permission we can just use psql locally. Is there some kind of "permission" advantage to this console or are you saying this console will be superior to psql to investigate the postgres database?

**Comment by yoomlam** [comment]
@au2 Good question!
This console is motivated by my pleasant experience with using the Rails console and its ORM capabilities. While someone could run SQL commands, it’s much easier to use the ORM models and functions we already have in the code base, which this console has access to. As our DB schema becomes more complicated, retrieving data across tables will be a lot easier with ORM than using JOINs (and less error prone). 
Plus since the console is in a Java environment, we have access to all the classes so we can test against real production data to diagnose particularly different bugs that can’t be replicated elsewhere. We’ll also have access to RabbitMQ, where we could replay messages to resume processing of previously failing claim (after a fix is deployed). 

I googled to see if there is a similar tool for Java and found the closest was Grails but it requires it’s own ORM model — I want to use the ORM we already have.

File: app/src/docker/docker-compose.yml [hunk 6]
```diff
@@ -182,3 +183,29 @@ services:
...
+      # TODO: move these Postgres and RabbitMQ configurations into a reusable yaml block
+      POSTGRES_HOST: vro-api-postgres-service
+      POSTGRES_DB: ${POSTGRES_DB}
+      POSTGRES_USER: ${POSTGRES_USER}
+      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
+      POSTGRES_URL: ${POSTGRES_URL} #jdbc:postgresql://vro-api-postgres-service:5432/vro
+      POSTGRES_SCHEMA: claims
+      RABBITMQ_PLACEHOLDERS_HOST: vro-api-rabbit-mq-service
+      RABBITMQ_PLACEHOLDERS_USERNAME: ${RABBITMQ_PLACEHOLDERS_USERNAME}
+      RABBITMQ_PLACEHOLDERS_USERPASSWORD: ${RABBITMQ_PLACEHOLDERS_USERPASSWORD}
```
Comment (au2) [inline app/src/docker/docker-compose.yml:201]: Needs to use the reusable block.

Comment (yoomlam) [reply app/src/docker/docker-compose.yml:201]: Yup, the reusable blocks PR wasn't merged yet. Updating now.

Comment (au2) [reply app/src/docker/docker-compose.yml]: @yoomlam Can you also correct names in vro-console depends-on block?

File: app/src/docker/docker-compose.yml [hunk 7]
```diff
@@ -182,3 +186,21 @@ services:
...
+    tty: true
+    environment:
+      <<: *redis-placeholder-vars
+      <<: *rabbitmq-placeholder-vars
+      <<: *postgres-vars
+    depends_on:
+      postgres-service:
+        condition: service_started
+      rabbitmq-service:
+        condition: service_healthy
```
Comment (yoomlam) [inline app/src/docker/docker-compose.yml:204]: Updated to use correct container names.

**Comment by yoomlam** [comment]
@au2 You should also get a `groovy:000> ` prompt. Garbage input should produce `Unknown property`:
```
groovy:000> asdf
Unknown property: asdf
groovy:000> ?

For information about Groovy, visit:
    http://groovy-lang.org 

Available commands:
  :help      (:h ) Display this help message
  ?          (:? ) Alias to: :help
  :exit      (:x ) Exit the shell
...
```

I don't think the container is receiving your stdin. Note this:
https://github.com/department-of-veterans-affairs/abd-vro/blob/f265990cf23b648445d37425c156b118bb7c754c/app/src/docker/docker-compose.yml#L193-L195

**Comment by yoomlam** [comment]
@au2 Are there any errors in the log: `docker logs vro-console-1`? Here's what to expect:

```
❯ docker logs vro-console-1

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::                (v2.7.4)

2022-10-20 16:53:42.419  INFO 1 --- [           main] g.v.v.c.ConsoleGroovyApplication         : Starting ConsoleGroovyApplication using Java 17.0.4.1 on 5cb5f3bae308 with PID 1 (/project/vro-console.jar started by docker in /project)
2022-10-20 16:53:42.456  INFO 1 --- [           main] g.v.v.c.ConsoleGroovyApplication         : No active profile set, falling back to 1 default profile: "default"
2022-10-20 16:53:49.224  INFO 1 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
2022-10-20 16:53:49.859  INFO 1 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 566 ms. Found 2 JPA repository interfaces.
2022-10-20 16:53:59.994  INFO 1 --- [           main] o.hibernate.jpa.internal.util.LogHelper  : HHH000204: Processing PersistenceUnitInfo [name: default]
2022-10-20 16:54:00.864  INFO 1 --- [           main] org.hibernate.Version                    : HHH000412: Hibernate ORM core version 5.6.11.Final
2022-10-20 16:54:02.797  INFO 1 --- [           main] o.hibernate.annotations.common.Version   : HCANN000001: Hibernate Commons Annotations {5.1.2.Final}
2022-10-20 16:54:03.300  INFO 1 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Starting...
2022-10-20 16:54:03.902  INFO 1 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Start completed.
2022-10-20 16:54:04.002  INFO 1 --- [           main] org.hibernate.dialect.Dialect            : HHH000400: Using dialect: org.hibernate.dialect.PostgreSQL10Dialect
2022-10-20 16:54:10.160  INFO 1 --- [           main] o.h.e.t.j.p.i.JtaPlatformInitiator       : HHH000490: Using JtaPlatform implementation: [org.hibernate.engine.transaction.jta.platform.internal.NoJtaPlatform]
2022-10-20 16:54:10.212  INFO 1 --- [           main] j.LocalContainerEntityManagerFactoryBean : Initialized JPA EntityManagerFactory for persistence unit 'default'
2022-10-20 16:54:21.246  INFO 1 --- [           main] o.a.c.impl.engine.AbstractCamelContext   : Apache Camel 3.19.0 (camel-1) is starting
2022-10-20 16:54:21.385  INFO 1 --- [           main] o.a.c.impl.engine.AbstractCamelContext   : Routes startup (started:1)
2022-10-20 16:54:21.391  INFO 1 --- [           main] o.a.c.impl.engine.AbstractCamelContext   :     Started route1 (seda://foo)
2022-10-20 16:54:21.393  INFO 1 --- [           main] o.a.c.impl.engine.AbstractCamelContext   : Apache Camel 3.19.0 (camel-1) started in 1s617ms (build:273ms init:1s214ms start:130ms)
2022-10-20 16:54:21.448  INFO 1 --- [           main] g.v.v.c.ConsoleGroovyApplication         : Started ConsoleGroovyApplication in 44.475 seconds (JVM running for58.953)
Working Directory = /project
2022-10-20 16:54:21.714  INFO 1 --- [           main] java.util.prefs                          : Created user preferences directory.
Running groovysh.rc
Variables:
  claimsT = org.springframework.data.jpa.repository.support.SimpleJpaRepository@173511ff
  vetT = org.springframework.data.jpa.repository.support.SimpleJpaRepository@53cddaf8
  camel = SpringCamelContext(camel-1) with spring id application
  pt = org.apache.camel.impl.engine.DefaultProducerTemplate@4bd51d3e
  _ = groovysh_evaluate$_run_closure1@d6b532f
  getAnyClaim = groovysh_evaluate$_run_closure1@d6b532f
Groovy Shell (3.0.13, JVM: 17.0.4.1)
Type ':help' or ':h' for help.
-------------------------------------------------------------------------------
groovy:000> asdf
Unknown property: asdf
```

**Review by au2 — APPROVED** [review]
I shut the terminal down and restarted and now it is working. Not sure what the problem before; I had the same steps.

It works. I was able to get the claimSubmissionId from a claim entity. I think I understand the value now.
